source/common/router/upstream_request.h

```diff
@@ -144,5 +147,64 @@ class UpstreamRequest : public Logger::Loggable<Logger::Id::router>,
   bool record_timeout_budget_ : 1;
 };
 
+// A generic API which covers common functionality between HTTP and TCP upstreams.
+class GenericUpstream {
+public:
+  virtual ~GenericUpstream() {}
+  virtual void encodeData(Buffer::Instance& data, bool end_stream) PURE;
+  virtual void encodeMetadata(const Http::MetadataMapVector& metadata_map_vector) PURE;
+  virtual void encodeHeaders(const Http::RequestHeaderMap& headers, bool end_stream) PURE;
+  virtual void encodeTrailers(const Http::RequestTrailerMap& trailers) PURE;
+  virtual void readDisable(bool disable) PURE;
+  virtual void resetStream() PURE;
+};
+
+class HttpUpstream : public GenericUpstream, public Http::StreamCallbacks {
+public:
+  HttpUpstream(UpstreamRequest& upstream_request, Http::RequestEncoder* encoder)
+      : upstream_request_(upstream_request), request_encoder_(encoder) {
+    request_encoder_->getStream().addCallbacks(*this);
```

Comment:
does this mean that the router no longer needs to implement `Http::StreamCallbacks`?

Reply:
good catch!  I lost that cleanup on rebase.